es74x-cloud/src/main/scala/tech/beshu/ror/proxy/es/ProxyIndexLevelActionFilter.scala

```diff
@@ -0,0 +1,115 @@
+import tech.beshu.ror.es.request.usermetadata.CurrentUserMetadataRequestHandler
+import tech.beshu.ror.proxy.es.ProxyIndexLevelActionFilter.ThreadRepoChannelRenewalOnChainProceed
+import tech.beshu.ror.proxy.es.clients.RestHighLevelClientAdapter
+import tech.beshu.ror.proxy.es.providers.{EsRestClientBasedRorClusterService, ProxyAuditSink, ProxyIndexJsonContentManager}
+
+import scala.util.{Failure, Success, Try}
+
+class ProxyIndexLevelActionFilter private(rorInstance: RorInstance,
+                                          esClient: RestHighLevelClientAdapter,
+                                          threadPool: ThreadPool)
+                                         (implicit scheduler: Scheduler)
+  extends ActionFilter {
+
+  private val rorClusterService = new EsRestClientBasedRorClusterService(esClient)
+
+  override def order(): Int = 0
+
+  override def apply[Request <: ActionRequest, Response <: ActionResponse](task: Task,
+                                                                           action: String,
+                                                                           request: Request,
+                                                                           listener: ActionListener[Response],
+                                                                           chain: ActionFilterChain[Request, Response]): Unit = {
+    (rorInstance.engine, ProxyThreadRepo.getRestChannel) match {
+      case (Some(engine), Some(channel)) =>
+        Try {
+          handleRequest(
+            engine,
+            task,
+            action,
+            request,
+            listener.asInstanceOf[ActionListener[ActionResponse]],
+            new ThreadRepoChannelRenewalOnChainProceed(chain, channel).asInstanceOf[ActionFilterChain[ActionRequest, ActionResponse]],
+            channel
+          )
+        } match {
+          case Success(_) =>
+          case Failure(ex) =>
+            listener.onFailure(new SecurityPermissionException("Cannot handle proxy request", ex))
+        }
+      case (Some(_), None) =>
+        chain.proceed(task, action, request, listener)
+      case (None, Some(channel)) =>
+        channel.sendResponse(createRorNotReadyYetResponse(channel))
+      case (None, None) =>
+        throw new IllegalStateException("Cannot process current request")
+    }
+  }
+
+  def stop(): MTask[Unit] = rorInstance.stop()
+
+  private def handleRequest(engine: Engine,
+                            task: Task,
+                            action: String,
+                            request: ActionRequest,
+                            listener: ActionListener[ActionResponse],
+                            chain: ActionFilterChain[ActionRequest, ActionResponse],
+                            channel: RestChannel): Unit = {
+    val requestInfo = new RequestInfo(channel, task.getId, action, request, rorClusterService, threadPool, false)
+    val requestContext = EsRequestContext.from(requestInfo).get
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Usage of get on optional type.](https://app.codacy.com/app/scarduzio/elasticsearch-readonlyrest-plugin/pullRequest?prid=4591272)